addons/sourcemod/scripting/nd_discord_playerlist.sp

```diff
@@ -0,0 +1,192 @@
+Handle SaveTeamsTimer;
+bool SavingTeams;
+Handle DatabaseHandle;
+
+public Plugin myinfo =
+{
+    name        = "[ND] Discord Player List",
+    author      = "zookatron",
+    description = "Provides player list information to Discord",
+    version     = "dummy",
+    url         = "https://github.com/stickz/Redstone/"
+};
+
+public void OnPluginStart()
+{
+    if (!SQL_CheckConfig("discord"))
+    {
+        SetFailState("Could not find configuration for \"discord\" database");
+        return;
+    }
+
+    SQL_TConnect(OnDatabaseConnected, "discord");
+
+    HookEvent("player_team", Event_PlayerTeam);
+}
+
+public Action Event_PlayerTeam(Event event, const char[] name, bool dontBroadcast)
+{
+    SaveTeams();
+    return Plugin_Continue;
+}
+
+public void OnClientPutInServer(int client)
+{
+    SaveTeams();
+}
+
+public void OnClientDisconnect_Post(int client)
+{
+    SaveTeams();
+}
+
+public OnDatabaseConnected(Handle owner, Handle handle, const char[] error, any:data)
+{
+    if (!IsValidHandle(handle))
+    {
+        SetFailState("Unable to connect to \"discord\" database: %s", error);
+        return;
+    }
+
+    DatabaseHandle = handle;
+
+    char query[1024];
+    FormatEx(query, sizeof(query), "SELECT * FROM players;");
+    SQL_TQuery(DatabaseHandle, Setup_Step1_CheckTable, query);
+}
+
+void Setup_Step1_CheckTable(Handle owner, Handle handle, const char[] error, any:data)
+{
+    // Create the table if it doesn't exist
+    if (error[0] != '\0')
+    {
+        char query[1024];
+        FormatEx(query, sizeof(query), "CREATE TABLE `players` (`id` bigint NOT NULL AUTO_INCREMENT, `name` varchar(128) NOT NULL, `team` varchar(128) NOT NULL, PRIMARY KEY (`id`));");
+        SQL_TQuery(DatabaseHandle, Setup_Step2_CreateTable, query);
+    }
+}
+
+void Setup_Step2_CreateTable(Handle owner, Handle handle, const char[] error, any:data)
+{
+    if (error[0] != '\0')
+    {
+        SetFailState("Unable to create \"players\" table in \"discord\" database: %s", error);
+    }
+}
+
+void SaveTeams()
+{
+    delete SaveTeamsTimer;
+    SaveTeamsTimer = CreateTimer(1.0, Timer_SaveTeams);
+}
```

Comment:
`SaveTeams()` should not fire unless the round is started. Best practice is to empty the table and populate it when the round starts. See [nd_rounds.inc](https://github.com/stickz/Redstone/blob/master/addons/sourcemod/scripting/include/nd_rounds.inc) for functionality.

Reply:
I would really like to not have to do this, this makes it so the player list is not accurate between rounds. In the game you are able to press tab to see an accurate list of players connected even between rounds and I would like this to do that as well. Do you really think it is a problem to track the players between rounds?

Reply:
Yes, it's a huge problem. It will spam queries when we shuffle teams. Instead, you can save teams pre-round by listing everyone as spectator. You can accomplish this with [OnClientConnected](https://sm.alliedmods.net/new-api/clients/OnClientConnected). Check the boolean value `ND_RoundStarted()`. Then when the round is started call `SaveTeams()` once on `public void ND_OnRoundStarted()`.

Reply:
I don't understand what you mean by spam queries, the code only saves anything if no changes have happened for one second so a team shuffle would only generate one query.

As far as I can tell I'm already doing exactly what you asked regarding saving teams pre-round by listing everyone as spectator, that's what it currently does. However that will stop working if I'm not allowed to call SaveTeams between rounds, which is what you asked me to do here? How is it possible for me to save the teams without saving the teams?

Reply:
One option would be to run a single query between when round_end (ND_OnRoundEnded) and round_start (ND_OnRoundStarted) occurs to `UPDATE players SET team = 'spectator'`

Reply:
> One option would be to run a single query between when round_end (ND_OnRoundEnded) and round_start (ND_OnRoundStarted) occurs to `UPDATE players SET team = 'spectator'`

This would cause the player list to be inaccurate during the time between rounds if people joined or quit in between rounds which is quite common. Again, there doesn't seem to be any need for worrying about this at all as far as I can see as there are no performance issues with the plugin tracking players between rounds, it does at the absolute most one low priority query per second (and usually far less than that) which is a virtually undetectable amount of overhead.